cmd/flat-network-controller/app/controller.go

```diff
@@ -0,0 +1,247 @@
+		klog.Warningf("failed to retrieve network in local cache by tenant %s, name %s: %v", tenant, name, err)
+		c.queue.Forget(item)
+		return
+	}
+
+	if net.Spec.Type != "flat" {
+		klog.V(5).Infof("network %s/%s is of type %q; ignored", net.Tenant, net.Name, net.Spec.Type)
+		c.queue.Forget(item)
+		return
+	}
+
+	klog.V(5).Infof("processing network %s/%s", net.Tenant, net.Name)
+
+	if err := manageFlatNetwork(net, c.netClientset, c.svcClientset); err != nil {
+		c.queue.AddRateLimited(key)
+		c.recorder.Eventf(net, corev1.EventTypeWarning, "FailedProvision", "failed to provision network %s/%s/%s: %v", net.Tenant, net.Tenant, net.Name, err)
+		return
+	}
+
+	c.recorder.Eventf(net, corev1.EventTypeNormal, "SuccessfulProvision", "successfully provision network %s/%s/%s", net.Tenant, net.Tenant, net.Name)
+	c.queue.Forget(item)
+}
+
+// manageFlatNetwork is the core logic to manage a flat typed network object
+func manageFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface) error {
+	if len(net.Status.DNSServiceIP) != 0 {
+		klog.V(5).Infof("network %s/%s/%s already have DNS service IP %s; skipped", net.Tenant, net.Namespace, net.Name, net.Status.DNSServiceIP)
+		return nil
+	}
+
+	svc, err := createOrGetDNSService(net, svcClient)
+	if err != nil {
+		return err
+	}
+
+	// since dns service is in place, flat type network is Ready now
+	netReady := net.DeepCopy()
+	netReady.Status.Phase = v1.NetworkReady
+	netReady.Status.DNSServiceIP = svc.Spec.ClusterIP
+	netReady.Status.Message = "DNS service ready; network ready"
+	_, err = netClient.ArktosV1().NetworksWithMultiTenancy(netReady.Tenant).UpdateStatus(netReady)
+	return err
+}
+
+func createOrGetDNSService(net *v1.Network, svcClient kubernetes.Interface) (*corev1.Service, error) {
+	const nsDNS = "kube-system"
+	nameDNS := "kube-dns-" + net.Name
+	dns := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nameDNS,
+			Tenant:    net.Tenant,
+			Namespace: nsDNS,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "dns",
+					Protocol:   "UDP",
+					Port:       53,
+					TargetPort: intstr.FromInt(53),
+				},
+				{
+					Name:       "dns-tcp",
+					Protocol:   "TCP",
+					Port:       53,
+					TargetPort: intstr.FromInt(53),
+				},
+				{
+					Name:       "metrics",
+					Protocol:   "TCP",
+					Port:       9153,
+					TargetPort: intstr.FromInt(9153),
+				},
+			},
+			Selector: map[string]string{
+				"k8s-app": "kube-dns",
+			},
+			Type: "ClusterIP",
+		},
+	}
+	svc, err := svcClient.CoreV1().ServicesWithMultiTenancy(nsDNS, net.Tenant).Create(&dns)
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			// todo: consider differing temporary errors from permanent errors
+			// todo: to fail the network provision in case of permanent errors
+			return nil, err
+		} else {
+			svc, err = svcClient.CoreV1().ServicesWithMultiTenancy(nsDNS, net.Tenant).Get(nameDNS, metav1.GetOptions{})
```

Comment:
Why we still need to check the existence of the service object when the Create() reports the object already exists?

Reply:
we need to get the service IP address